### ISSUE TITLE
Why do we need to take care of num_process in lr_scheduler?

### ISSUE THREAD

**Issue by Hannibal046** [huggingface/accelerate#1382]
Hi, I am a little bit confused about the behavior of lr_scheduler especially this line of code:
https://github.com/huggingface/accelerate/blob/995563fec9801308b3ddbbfaf4128730182eaa81/src/accelerate/scheduler.py#L73-L82

In my humble opinion, in the `DDP` mode, `gradient_accumulation_steps` and `more gpus` are essentially increase the batch size . Suppose we the batch_size parameter we give to dataloader is 10, we have 8 gpus, and gradient accumulation steps is 4. So the real batch size is 10\*8\*4 = 320. This is i believe when we should do one optimization step as well as one lr schedule step. But in the above line of code, it would do more than one lr schedule step.

Briefly speaking, we question is: what do we need to make lr_sceduler depend on the number of GPUs?

**Comment by Hannibal046** [comment]
a relevant problem is here: https://github.com/huggingface/transformers/issues/18436#issuecomment-1414290934

**Comment by sgugger** [comment]
That's because the lr_scheduler has been defined taking into account the unprepared dataloader which has the original batch size.

**Comment by Hannibal046** [comment]
But in my case(2 GPUs), the size of `train_dataloader` doesn't change after the `prepare` method. I believe the DDP sampler wouldn't change the the size of `train_dataloader`?

BTW, thanks for the prompt response!

**Comment by Hannibal046** [comment]
So, if i want to assign `warmup_steps` to 4000, and i want my `lr` to exactly peak at 4000. (no matter how many GPUs or batch size i set). What would you suggest to be the best practice?

**Comment by Hannibal046** [comment]
Sorry, it my mistake. The dataloader size does change after the prepare method.

Could you please explain some conceptual guides on how to set the warmup_steps in accelerate. Take this paper, [Attention is all you need](https://arxiv.org/pdf/1706.03762.pdf) as example. In this paper, it said the model was trained on 8GPUs and warmup_steps is 4000. So, i should give 4000\*8 to the scheduler function?
![image](https://user-images.githubusercontent.com/38466901/235758019-8d07d028-bdb0-455e-b5e7-9fbb8a62bdf2.png)

```python
from transformers import get_polynomial_decay_schedule_with_warmup,
lr_scheduler = get_polynomial_decay_schedule_with_warmup(
        optimizer, 
        num_warmup_steps = 4000*8*cfg.trainer.gradient_accumulation_steps,
        num_training_steps = total_steps
    )
```

**Comment by Hannibal046** [comment]
Hi, @sgugger , would you mind spare some time to check this?

**Comment by Hannibal046** [comment]
Hi @sgugger , really sorry for bothering you. I am new to `Accelerate` and the `prepare` method works like a magic black box for me  to handle multi GPUs (which it is intended to be). So the obvious solution of not preparing `lr_scheduler` is not so obvious for me.

For anyone who is seeking for fixed warmup steps in Accelerate, a simple workaround would be:
```python
lr_scheduler = get_scheduler()

# not prepare lr_scheduler

for batch in dataloader:
	with accelerator.accumulate(model):
		loss = compute_loss()
		if accelerator.sync_gradients and not accelerator.optimizer_step_was_skipped:
			lr_scheduler.step()
	
 
```
